### PR TITLE
Support creating a publication for specified list of schemas. Also allow creating a publication for specified list of tables with specific columns and specific rows using WHERE clause. #5868

### PR DESCRIPTION
**Pull request by anilsahoo20**

### CODE REVIEW

File: web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py [hunk 1]
```diff
@@ -381,24 +394,63 @@ def _fetch_properties(self, did, pbid):
...
+                    schema_name_sql)
+
+                snames_list = []
+
+                if len(snames_list_res['rows']) != 0:
+                    for sname_dict in snames_list_res['rows']:
+                        snames_list.append(sname_dict['sname'])
+
+                if len(snames_list) != 0:
+                    res['rows'][0]['pubschema'] = snames_list
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py:432]: just use a list comprehension here to assign `sname` array directly. remove all over if codes.

File: web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py [hunk 2]
```diff
@@ -691,6 +821,34 @@ def get_sql(self, data, pbid=None):
...
+        server group id, server id and database id.
+
+        Args:
+            gid: Server Group ID
+            sid: Server ID
+            did: Database ID
+        """
+        res = []
+        try:
+            view = SchemaDiffRegistry.get_node_view('schema')
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py:837]: we should not use `SchemaDiffRegistry` here. It is meant for using in SchemaDiff only.

File: web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py [hunk 3]
```diff
@@ -710,38 +868,97 @@ def get_tables(self, gid, sid, did):
...
+        if not status:
+            return internal_server_error(errormsg=res)
+
+        if self.manager.version < 150000:
+            for row in rset['rows']:
+                res.append(
+                    {
+                        'label': row['table'],
+                        'value': row['table']
+                    }
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py:880]: you can simply change the old SQL to add `tid` and make the code common instead of adding so many if conditions.

File: web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py [hunk 4]
```diff
@@ -710,38 +868,97 @@ def get_tables(self, gid, sid, did):
...
-
-        table_sql = render_template(
-            "/".join([self.template_path, self._GET_TABLE_FOR_PUBLICATION]),
-            pname=pname
-        )
+        if self.manager.version < 150000:
+            table_sql = render_template(
+                "/".join([self.template_path,
+                          self._GET_TABLE_FOR_PUBLICATION]),
+                pname=pname
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py:940]: Template data can be created based on conditions and then passed as `**data` to avoid code repeat.

File: web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py [hunk 5]
```diff
@@ -777,14 +994,31 @@ def sql(self, gid, sid, did, pbid, json_resp=True):
...
             return gone(self._NOT_FOUND_PUB_INFORMATION)
 
+        snames_list = []
+
+        if self.manager.version >= 150000:
+            schema_name_sql = render_template(
+                "/".join([self.template_path, 'get_schemas.sql']),
+                pbid=pbid
+            )
+            status, snames_list_res = self.conn.execute_dict(
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/__init__.py:1004]: I can see this code is repeated at many places. Can't we make it common?

File: web/pgadmin/browser/server_groups/servers/databases/publications/static/js/publication.ui.js [hunk 6]
```diff
@@ -33,31 +32,146 @@ export class DefaultWithSchema extends BaseUISchema {
...
+      msg = gettext('Supported version <= %s', variable?.max_server_version);
+    }
+
+    return msg;
+  }
+
+  getTableOid(tabName) {
+    // Here we will fetch the table oid from table name
+    // iterate over list to find table oid
+    // console.log(this.allTablesOptions)
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/static/js/publication.ui.js:79]: remove commented code.

File: web/pgadmin/browser/server_groups/servers/databases/publications/templates/publications/sql/15_plus/get_tables.sql [hunk 7]
```diff
@@ -0,0 +1,13 @@
+SELECT pg_catalog.quote_ident(n.nspname) || '.' || pg_catalog.quote_ident(cls.relname) AS TABLE_NAME,
+	(SELECT array_agg(attname)
+		FROM pg_attribute att
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/templates/publications/sql/15_plus/get_tables.sql:3]: please use formatter to align from and select.

File: web/pgadmin/browser/server_groups/servers/databases/publications/tests/15_plus/create_publication_msql.sql [hunk 8]
```diff
@@ -0,0 +1,4 @@
+CREATE PUBLICATION test_publication_create
+   FOR ALL TABLES
+
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/tests/15_plus/create_publication_msql.sql:3]: Why empty line?

File: web/pgadmin/browser/server_groups/servers/databases/publications/tests/test_publication_put.py [hunk 9]
```diff
@@ -86,16 +106,181 @@ def runTest(self):
...
+            else:
+                tables = publication_utils.get_tables(self)
+                self.test_data['pubtable'] = [tables[0]]
+
+        if self.server_version >= 150000:
+            if hasattr(self, 'add_schema'):
+                self.test_data['pubtable'] = ''
+                self.test_data['pubschema'] = \
+                    publication_utils.get_schemas(self)
+            if hasattr(self, 'add_table_with_where'):
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/tests/test_publication_put.py:144]: Don't put conditions this way. Use scenarios instead of so many if conditions. Code should behave based on schenarios.

File: web/pgadmin/browser/server_groups/servers/databases/publications/tests/utils.py [hunk 10]
```diff
@@ -27,7 +27,48 @@ def get_tables(self):
             self.server_id) + '/' +
         str(self.db_id) + '/',
         content_type='html/json')
-    return [tables.json['data'][0]['value']]
+    if self.server_version >= 150000:
+        columns = get_all_columns(self)
+        if hasattr(self, 'with_columns'):
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/tests/utils.py:32]: again.return data is hardcoded here. it should be based on scenario.

File: web/pgadmin/static/js/SchemaView/MappedControl.jsx [hunk 11]
```diff
@@ -55,7 +55,7 @@ function MappedFormControlBase({ type, value, id, onChange, className, visible,
   case 'password':
     return <FormInputText name={name} value={value} onChange={onTextChange} className={className} type='password' inputRef={inputRef} {...props} />;
   case 'select':
-    return <FormInputSelect name={name} value={value} onChange={onTextChange} className={className} inputRef={inputRef} {...props} />;
+    return <FormInputSelect name={name} value={value == ''? []:value} onChange={onTextChange} className={className} inputRef={inputRef} {...props} />;
```
Comment (adityatoshniwal) [inline web/pgadmin/static/js/SchemaView/MappedControl.jsx:58]: why is this changed?

**Review by adityatoshniwal — CHANGES_REQUESTED** [review]
I have done a basic code review. Not tested though.

File: web/pgadmin/static/js/SchemaView/MappedControl.jsx (hunk 11 above)
Comment (anilsahoo20) [reply web/pgadmin/static/js/SchemaView/MappedControl.jsx:58]: The multi select dropdown was giving an empty option when cleared, So added that. But currently looking for a workaround on the publication code itself.

File: web/pgadmin/browser/server_groups/servers/databases/publications/tests/15_plus/create_publication_msql.sql (hunk 8 above)
Comment (anilsahoo20) [reply web/pgadmin/browser/server_groups/servers/databases/publications/tests/15_plus/create_publication_msql.sql:3]: The sql empty line gets generated if all_tables switch is made to 'Yes' as the pubschema and pubtable will not be present in the data. Will fix try to fix this as well.

File: web/pgadmin/static/js/SchemaView/MappedControl.jsx (hunk 11 above)
Comment (anilsahoo20) [reply web/pgadmin/static/js/SchemaView/MappedControl.jsx]: Resolved

File: web/pgadmin/browser/server_groups/servers/databases/publications/tests/15_plus/create_publication_msql.sql (hunk 8 above)
Comment (anilsahoo20) [reply web/pgadmin/browser/server_groups/servers/databases/publications/tests/15_plus/create_publication_msql.sql]: Resolved

File: web/pgadmin/browser/server_groups/servers/databases/publications/static/js/publication.ui.js [hunk 12]
```diff
@@ -33,31 +32,145 @@ export class DefaultWithSchema extends BaseUISchema {
...
+      {
+        id: 'where',
+        label: gettext('Where'),
+        type: 'sql',
+        deps: ['table_name'],
+        disabled: (state) => !obj.isTableName(state),
+        editable: function (state) {
+          return obj.isNew(state) || !obj.allReadOnly;
+        },
+        // schema: this.paramSchema,
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/static/js/publication.ui.js:148]: remove commented code.

File: web/pgadmin/browser/server_groups/servers/databases/publications/templates/publications/sql/15_plus/get_schemas.sql [hunk 13]
```diff
@@ -0,0 +1,4 @@
+SELECT n.nspname AS sname
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/templates/publications/sql/15_plus/get_schemas.sql:1]: rename file to get_pub_schemas.sql

File: web/pgadmin/browser/server_groups/servers/databases/publications/templates/publications/sql/15_plus/schema_based.sql [hunk 14]
```diff
@@ -0,0 +1,3 @@
+SELECT pnnspid AS sid
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/templates/publications/sql/15_plus/schema_based.sql:1]: isn't the get_schemas.sql and this file same?

File: web/pgadmin/browser/server_groups/servers/databases/publications/tests/test_publication_put_add_table.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+        self.test_data['id'] = self.publication_id
+
+        if self.is_positive_test:
+            if hasattr(self, 'add_table'):
+                self.table_name = "table_column_%s" % (str(uuid.uuid4())[1:8])
+                self.table_id = tables_utils.create_table(self.server,
+                                                          self.db_name,
+                                                          self.schema_name,
+                                                          self.table_name)
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/server_groups/servers/databases/publications/tests/test_publication_put_add_table.py:90]: This code is repeated for all the cases.

**Review by adityatoshniwal — CHANGES_REQUESTED** [review]
